### DISCUSSION TITLE
minipage .columns layout

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#4959] [Q&A]
Maybe I've missed something, but I can't seem to manage getting a multi-column layout in pdf format. I think to get the same functionality as in [HTML presentations](https://quarto.org/docs/presentations/index.html#multiple-columns) or even [Beamer](https://quarto.org/docs/presentations/beamer.html#multiple-columns), one would need to use `minipage` environments, which is [what beamer uses internally](https://github.com/josephwright/beamer/blob/772931f0da783b724a255235ad6935755afdc439/base/beamerbaseframecomponents.sty#L193) in its definition of a columns environment.

It would be quite nice if the same syntax could work in other latex classes as well,

```

:::: {.columns}

::: {.column width="60%"}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::
::: {.column width="10%"}
:::
::: {.column width="30%"}
![](dummy.png)
:::

::::
```

![Screenshot 2023-03-25 at 20 30 21](https://user-images.githubusercontent.com/18970/227703733-ad7edba4-6063-4569-886b-393ccfe59759.png)

Currently, I'm resorting to raw LaTeX, but that means everything inside the `minipage` environment needs to be in latex, so that quickly defeats the purpose of authoring in quarto markdown in the first place. 
If one wants markdown input, then the raw latex commands need to be tagged as raw latex, which is arguably a bit nicer, but still clutters quite a lot the document (see "mixed" section below).

I've tried defining a custom div `::: minipage`, but the problem is that they're always followed by an extra blank line, which breaks the LaTeX layout (there cannot be any empty line between two minipages, if they are to be side-by-side).

I guess the ideal solution would be:

- to somehow extract the [beamer columns](https://github.com/josephwright/beamer/blob/772931f0da783b724a255235ad6935755afdc439/base/beamerbaseframecomponents.sty#L193) definition and add it to the latex template (looking at it, it's probably easier to write one from scratch, as it's full of beamer-specific _things_)
- or, if possible, avoid the trailing blank line after divs

Illustrative example below:

````
---
title: "mixed columns"
format: pdf
---

## Latex

\begin{minipage}[t]{.6\textwidth}
\vspace{0pt}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.
  \begin{enumerate}
    \item firstly
    \item secondly
    \item thirdly
    \item quarto
  \end{enumerate}
\end{minipage}
\hfill
\begin{minipage}[t]{.3\textwidth}
\vspace{0pt}
\includegraphics{dummy.png}
this is a picture
\end{minipage}

## quarto columns

:::: {.columns}

::: {.column width="60%"}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::
::: {.column width="10%"}
:::
::: {.column width="30%"}
![](dummy.png)
:::

::::

## mixed

```{=latex}
\begin{minipage}[t]{.6\textwidth}
\vspace{0pt}
```
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
  
```{=latex}
\end{minipage}
\hfill
\begin{minipage}[t]{.3\textwidth}
\vspace{0pt}
```
![](dummy.png)
this is a picture
```{=latex}
\end{minipage}
```
````

**Comment by baptiste** [comment]
I got some [help on TeX.stackexchange](https://tex.stackexchange.com/a/680753/269911) for the latex side of things – it turns out there's a beamerarticle package which defines beamer macros for use in other classes:

````
\documentclass{article}

\usepackage{graphicx}
\usepackage{beamerarticle}

\makeatletter
\mode<article>{\renewenvironment<>{beamer@columnenv}[2][]{ \begin{minipage}[t]{#2}}{\end{minipage}}
\let\endcolumn\endbeamer@columnenv}
\makeatother

\begin{document}

\begin{columns}
\begin{column}{.6\textwidth}
\vspace{0pt}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.
  \begin{enumerate}
    \item firstly
    \item secondly
    \item thirdly
    \item quarto
  \end{enumerate}
\end{column}%
\hfill
\begin{column}{.3\textwidth}
\vspace{0pt}
\includegraphics[width=\textwidth]{dummy.png}
this is a picture
\end{column}%
\end{columns}

\end{document}
````

At this stage, I thought I could use the `latex-environment` extension to test this in quarto,

````
---
title: "mixed columns"
filters:
   - latex-environment
environments: [columns, column]
format: 
  pdf:
    keep-tex: true
    include-in-header: 
      text: |
        \usepackage{beamerarticle}
        
        \makeatletter
        \mode<article>{\renewenvironment<>{beamer@columnenv}[2][]{ \begin{minipage}[t]{#2}}{\end{minipage}}
        \let\endcolumn\endbeamer@columnenv}
        \makeatother
---


:::: {.columns}

::: {.column arguments=".6\textwidth"}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::
::: {.column arguments=".3\textwidth"}
![](dummy.png)
:::

::::
````

it _almost_ works, unfortunately, the extra blank line problem hits again, and the last minipage goes to the next line.

![Screenshot 2023-03-26 at 08 42 49](https://user-images.githubusercontent.com/18970/227738181-3ce59659-a3b0-4a44-9962-a39106efdbab.png)

I looked for a Lua filter in quarto where the `.column` behaviour is defined for the various formats, but I didn't find it. I'm guessing it's implemented as a custom AST node?

I think this would be a very useful addition for pdf documents, matching the behaviour of other formats.

**Reply by samcarter** [reply]
The reason you get the image in a new line is that quarto/pandoc/whatever seems to insert empty lines before and after all the `\begin`s and `\end`s of the environments, which will instruct latex to insert all these undesired line breaks

**Reply by baptiste** [reply]
Yep, it's been reported several times, but I'm not sure there's an easy fix.

**Reply by samcarter** [reply]
The code produced from your example also lacks `%` at many line endings which will result in overfull boxes.

**Reply by baptiste** [reply]
ah, thanks, I've never quite understood `%` at line endings (or worried too much about LaTeX warning me about bad boxes ;)

**Reply by samcarter** [reply]
... I wonder: if you have this `latex-environment` extension to create new filters, why don't you create one for minipages instead of columns? This would work with all classes.

**Reply by samcarter** [reply]
Something like this
```
---
title: "mixed columns"
filters:
   - latex-environment
environments: [minipage]
format: 
  pdf
---

::: {.minipage arguments=".6\textwidth"}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::
::: {.minipage arguments=".3\textwidth"}
![](example-image.png)
:::
```

(the empty lines are still wrong and the `%` are missing, but I suppose that's something you should report to the `latex-environment` extension )

**Reply by baptiste** [reply]
I tried this first, and it almost works, but the extra line breaks the layout. Sadly, the trailing line is a widespread issue and as far as I remember is due to the way pandoc processes AST nodes, so not that easy to fix as it's kind of intentional and downstream.

**Comment by mcanouil** [comment]
First your code is not fully reproducible.

This being said, the "layout" multiple columns syntax works just fine.
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/227739313-69ae1ef3-6ab7-4a17-8ad2-e36900bed985.png">
````qmd
---
title: "quarto-playground"
format: pdf
---

::: {layout="[ 0.6, -0.1, 0.3 ]"}

:::: {#first-column}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::

::: {#second-column}
```{r}
plot(1)
```
:::

::::
````

---

### `layout-ncol="2"`

````qmd
---
title: "quarto-playground"
format: pdf
---

::: {layout-ncol="2"}

:::: {#first-column}
Lorem ipsum dolor sit amet consectetur adipiscing elit morbi, vivamus erat himenaeos litora cras magnis blandit leo, etiam turpis lectus nullam feugiat porttitor dis. Non ultricies mus turpis in habitant auctor euismod duis scelerisque porttitor, mattis metus nisl risus leo curabitur laoreet eget varius. Etiam sed tortor nulla quis vehicula sodales non convallis aliquam phasellus, ut facilisis sociis molestie vel sagittis eu luctus.

1. firstly
1. secondly
1. thirdly
1. quarto
:::

::: {#second-column}
```{r}
plot(1)
```
:::

::::
````

**Reply by mcanouil** [reply]
PS: @baptiste Could you change the category of the discussion from "general" to "Q&A", as it feels more appropriate?

**Reply by baptiste** [reply]
Ah, nice! I wonder why `layout` works but `columns` isn't implemented as in the other formats. It's good to have a workaround for now! In the future, I hope it'll be possible to use the `::: {.column width="60%"}` syntax for simple situations like this, as it's easier on the eyes than `layout ="[ 0.6, -0.1, 0.3 ]"`

**Reply by mcanouil** [reply]
The `columns` / `column` syntax is quite limited and can't really be used for all cases.

`layout="[ 0.6, -0.1, 0.3 ]"` is to mimic your first example which was 60% 10% 30% for columns.
If you don't need spacing `layout="[ 0.6, 0.4 ]"`.
With the `layout` syntax you can build grids, _e.g._, `layout="[ [0.6, 0.4], [0.4, 0.6] ]"`.
<img width="1624" alt="image" src="https://user-images.githubusercontent.com/8896044/227740045-a32e8aa5-81e1-4a35-9dbd-5b4f90844af1.png">

**Reply by rainer-rq-koelle** [reply]
I think it would be useful to have the option to switch to a (nicely looking) 2-column layout for pdf output. I am not fully sure that I understand the `-value` for creating a gap between the columns. As you can see from the example above, text is center-aligned. If one uses two text blocks next to each other, the gap disappears in pdf output (unless set with a negative value, e.g. `layout="[0.35, -1, 0.55]"`. While it works somehow for me, the syntax could be something like `layout="[0.4, 0.6], gap = 0.1"` to make it more intuitive 

P.S. it took me a while to figure out that, if one uses a negative min-value to force a gap between the columns, there is no need for a separate div to account for this 'neg'-column. 
```
:::::: {layout="[0.35, -0.1, 0.55]"}
::: {#rightcolumn}
some cool stuff for the right column
:::
::: {#leftcolumn}
ahhhhh ... I do not need a dummy `::: <<blank>> :::` block for the -0.1 pseudo-column
::: 
:::::: 
```

P.S.S. I also noted that if one combines figures with text, there is a fair chance that the div is moved -- following LATEX rules -- to the next/another page. This behaviour is particularly popping up when one uses multiple rows (e.g. like the example above with surrounding text. If one spreads out every row into its separate 2-column block, likelihood is high that one block is moved to another page. Forcing `figure-pos: "H"` has little to ne effect on this behaviour.
Could anyone help how to handle a multi-row 2-column "insertion" for a pdf output?

**Reply by mcanouil** [reply]
In the YAML frontmatter you can set the "twocolumn" class option.
There is also the Pandoc columns/column: https://quarto.org/docs/presentations/beamer.html#multiple-columns
And of course you have access to all LaTeX via raw code blocks (https://quarto.org/docs/authoring/markdown-basics.html#raw-content).

Regarding the negative value, that's documented in https://quarto.org/docs/authoring/figures.html#complex-layouts

**Reply by mcanouil** [reply]
> Could anyone help how to handle a multi-row 2-column "insertion" for a pdf output?

@rainer-rq-koelle Please open a new discussion and follow the instructions/guidelines in the template.
